src/AWS.Logger.AspNetCore/AWSLogger.cs

```diff
@@ -74,49 +118,82 @@ public void Log<TState>(LogLevel logLevel, EventId eventId, TState state, Except
             {
                 if (formatter == null)
                     throw new ArgumentNullException(nameof(formatter));
-                message = formatter(state, exception);
-            }
-            else
-            {
-                message = _customFormatter(logLevel, state, exception);
-            }
 
-            if (string.IsNullOrEmpty(message) && exception == null)
-                // If neither a message nor an exception are provided, don't log anything (there's nothing to log, after all).
-                return;
+                // Format of the logged text, optional components are in {}
+                //  {[LogLevel] }{ => Scopes : }{Category: }{EventId: }MessageText {Exception}{\n}
+                var components = new List<string>(4);
+                if (IncludeLogLevel)
+                {
+                    components.Add($"[{logLevel}]");
+                }
 
-            var scope = IncludeScopes ? AddScopeInformation() : string.Empty;
-            if (exception != null && _customFormatter==null)
-            {
-                message = string.Concat(scope, message, Environment.NewLine, exception.ToString(), Environment.NewLine);
+                GetScopeInformation(components);
+
+                if (IncludeCategory)
+                {
+                    components.Add($"{_categoryName}:");
+                }
+                if (IncludeEventId)
+                {
+                    components.Add($"[{eventId}]:");
+                }
+
+                string text;
+                if (_customFormatter == null)
+                {
+                    text = formatter(state, exception);
+                }
+                else
+                {
+                    text = _customFormatter(logLevel, state, exception);
+                }
+
+                components.Add(text);
+
+                if (IncludeException)
+                {
+                    components.Add($"{exception}");
+                }
+                if (IncludeNewline)
+                {
+                    components.Add(Environment.NewLine);
+                }
+
+                message = string.Join(" ", components);
             }
             else
             {
-                message = string.Concat(scope, message, Environment.NewLine);
+                message = _customFormatter(logLevel, state, exception);
             }
+
             _core.AddMessage(message);
         }
 
-        private static string AddScopeInformation()
+        private void GetScopeInformation(List<string> logMessageComponents)
```

Comment:
I would rather cut down the amount of allocations done instead of combining multiple collections or StringBuilders later. If the method was public I would agree with you.